oxidation/libparsec/crates/types/src/error.rs

```diff
@@ -40,10 +40,10 @@ pub enum DataError {
     UnexpectedVersion { expected: u32, got: u32 },
 }
 
-pub type DataResult<T> = Result<T, DataError>;
+pub type DataResult<T> = Result<T, Box<DataError>>;
```

Comment:
What's the choice to box the error ?

Reply:
The size of the error, due to `UnexpectedAuthorError`

Reply:
How 2 `DeviceID` be too big ?

Reply:
Can't we just box the incriminated variant ?

Reply:
Is it better ?

Reply:
For me yes, it would require to do the heap allocation only for this variant

Reply:
Performance isn't needed when you catch errors, we should priorize the easiest way

Reply:
ok then

Reply:
I don't understand the need for Boxing error here :/

As @TimeEngineer said, "we should priorize the easiest way": so no Box at all is the best.

My understanding is boxing error is useful to return different types that are checked at runtime:
```rust
type Result<T> = std::result::Result<T, Box<dyn error::Error>>;
```

However this is not the case here, so what are we achieving that we can't without boxing here ?

Reply:
after some more digging:

- `String` [is literally a `Vec<u8>`](https://doc.rust-lang.org/src/alloc/string.rs.html#367-369)
- `Vec<T>` is composed of three words (number of elements in the vector, pointer of the buffer storing the elements and actual size of the buffer)
- So we endup on x86_64 with `Vec<T>`/`String` made of... 24bytes 😱  (I've checked using `std::mem::size_of` on Windows 10 64bits)
- Hence `DeviceID` weight 72bytes now that is contains 3 `String`... that quite heavy !

on the other hand, a box is 8 or 16bytes (depending if the [boxed element is a dynamically sized type](https://doc.rust-lang.org/reference/dynamically-sized-types.html))

So on paper it would be more efficient to use box...
![image](https://user-images.githubusercontent.com/3187637/209538438-3762964f-0200-4d54-9733-85391e3a9498.png)


We don't have actual proof this is the case.
- Using a box means we have to do a heap allocation (which is cheap, but less cheap than just using the stack).
- Rust relies heavily on compiler optimization & in-lining (hence the perf difference between dev and release profiles !), so it's very likely most function returning a `Result<T, MyBigAssErrorType>` gets optimized out and the result is only written once in it final location (hence making it similar in perf with the single write needed when initializing the box... less the heap allocation and pointer redirection during access ^^)

So in the end this use of Box [smells heavily like premature optimization](https://xkcd.com/1691/). I'd rather have simpler to use error (and not having to constantly ponder "do we need box for this new error type ?").

Reply:
edit: this has been changed due to [clippy codesmell](https://rust-lang.github.io/rust-clippy/master/index.html#result_large_err), I'm all okay to follow clippy Rules so carry on ! ^^